src/org/mozartoz/truffle/runtime/TailCallException.java

```diff
@@ -0,0 +1,17 @@
+package org.mozartoz.truffle.runtime;
+
+import com.oracle.truffle.api.nodes.ControlFlowException;
+
+public class TailCallException extends ControlFlowException {
+
+	private static final long serialVersionUID = -6432472909387052509L;
+
+	public final Object receiver;
+	public final Object[] arguments;
```

Comment:
Use `@CompilationFinal(dimensions=1)` here, it allows to fold reading the elements.